File: javascript/leetcode/count-complete-tree-nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var countNodes = function(root) {
    let count = 0;
    var traverse = function(root) {
        if (root) {
            count++;
            traverse(root.left);
            traverse(root.right);
        }
    }
    traverse(root);
    return count;
};

File: ruby/leetcode/count-complete-tree-nodes.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} count
# @return {Integer}
def traverse(root, count)
    new_count = count
    if root != nil
        new_count += 1
        new_count = traverse(root.left, new_count)
        new_count = traverse(root.right, new_count)
    end
    return new_count
end

# @param {TreeNode} root
# @return {Integer}
def count_nodes(root)
    return traverse(root, 0)
end

